website/hw/hw4.rst

```diff
@@ -180,103 +218,517 @@ As you are stepping through your program, you might want to be able to
+to know exactly where the segfault is happening.
+
+GDB is a really useful tool because it helps you pinpoint where a segfault is
 happening and what might be causing it. You can use all the commands
-we’ve already talked about, but there are a couple of additional ones
+we’ve already talked about (e.g., to step through a program line
+by line until it segfaults), but there are a couple of additional ones
 that might be useful.
 
-``kill`` will stop the current run of the program. You can then restart
-it by running ``run`` again. Frequently you’ll want to run a program
-until it segfaults, ``kill`` it (stop the program without exiting gdb),
-add breakpoints, and then run it again.
-
 ``backtrace`` will print out the call stack. What this means is that it
 will print the function being called as well as the function that called
-that, and the function that called that, etc.
+that, and the function that called that, etc. Using this command
+after a segfault in GDB can often tell us exactly what is causing the
+segfault.
 
 ``frame n`` (where ``n`` is the frame number from the backtrace) allows
 you to go to the function in that frame and see what line in that
 particular function caused the error.
 
-Compile ``task3.c`` (don’t forget to include the ``-g`` option!
-Otherwise you won’t be able to debug it). Now run it; the program should
-segfault. Now it’s time to debug! There are three errors in this file.
-For each error, answer the following question on Gradescope:
+If ``backtrace`` and ``frame n`` don't reveal any useful information,
+you may want to set breakpoints close to the point where the segfault happens,
+to see exactly what happens in the lines leading up to the segfault. However,
+to do this, you must first use the ``kill`` to stop the current run of the program
+(without exiting GDB).
+You will then be able to ``run`` the program again, stopping at any additional
+breakpoints you specify.
 
--  Where was the segfault? Give the function, line number, and
-   copy-paste the line of code itself. What was causing the segfault and
-   how did you fix it?
+**Practice Question E**: Try running the ``segfault`` program with gdb and using the above
+commands to figure out what is causing the segfault. You will find
+the answer at the bottom of this homework.
 
-It is a good idea to fix one error before moving on to the next. Make
-sure to also commit and push your fix for each segfault!
+Task 3: Fixing Runtime Errors in the Micro Editor
+-------------------------------------------------
 
-You must also answer the following question for the entire task:
+(40 points)
 
--  Describe the steps you took in gdb to narrow down the exact location
-   of one of the segfaults. You can also copy-paste the contents of one
-   debugging session with gdb, with a brief note at the end explaining
-   how that debugging session narrowed down the location of the error.
+We have inserted two bugs in the Micro code that will cause the editor to
```

Comment:
This error seems extremely difficult to find & fix

Reply:
Task 3.2 is meant to be a bit of a challenge question, and I plan to go over it in class after the homework is done. Maybe we should make it worth just 10 points so they don't feel bad if they can't figure it out?

Reply:
(also, we can give them partial credit for explaining how they tried to debug it, even if they didn't manage to fix the bug)

Reply:
That makes sense -- could we also add a little more documentation in the code itself? So that if they make it to identifying the line that's directly causing the problem, it's a little more clear what the line is trying to do, and so they can more quickly identify what the issue is. (In general that's the kind of code I think we'd want them to document anyway if they were writing it)